### DISCUSSION TITLE
How to use? Any tutorial on this?

### DISCUSSION THREAD

**Discussion by mrd33p** [lukas-blecher/LaTeX-OCR#71] [Q&A] [answered]
Hi,
I am not from a programming background, and I have no idea how to use this. I have been using Mathpix before this, but the free version runs out most of the time. I tried to follow the readme file but when I entered the command "pip install -r requirements.txt," it installed a few things, then it errored out with the following.

`ERROR: Command errored out with exit status 1:
     command: 'e:\ProgramData\Anaconda3\python.exe' -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\DELL\\AppData\\Local\\Temp\\pip-install-yt10_hyi\\entmax_2addef42568d45e9bd5ef7a110959299\\setup.py'"'"'; __file__='"'"'C:\\Users\\DELL\\AppData\\Local\\Temp\\pip-install-yt10_hyi\\entmax_2addef42568d45e9bd5ef7a110959299\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base 'C:\Users\DELL\AppData\Local\Temp\pip-pip-egg-info-4nr6qetp'
         cwd: C:\Users\DELL\AppData\Local\Temp\pip-install-yt10_hyi\entmax_2addef42568d45e9bd5ef7a110959299\
    Complete output (9 lines):
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "C:\Users\DELL\AppData\Local\Temp\pip-install-yt10_hyi\entmax_2addef42568d45e9bd5ef7a110959299\setup.py", line 2, in <module>
        from entmax import __version__
      File "C:\Users\DELL\AppData\Local\Temp\pip-install-yt10_hyi\entmax_2addef42568d45e9bd5ef7a110959299\entmax\__init__.py", line 3, in <module>
        from entmax.activations import sparsemax, entmax15, Sparsemax, Entmax15
      File "C:\Users\DELL\AppData\Local\Temp\pip-install-yt10_hyi\entmax_2addef42568d45e9bd5ef7a110959299\entmax\activations.py", line 13, in <module>
        import torch
    ModuleNotFoundError: No module named 'torch'
    ----------------------------------------
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.`

I have anaconda installed and probably the Latest version of python as well. When I tried running other .py commands, nothing happened. 
Please let me know what's causing the error, or the best would be to redirect me to any tutorial (if there is any) because I am not confident using this and probably will face more errors. 

Thank you for your work.

**Comment by lukas-blecher** [comment, marked as answer]
The reason is that the deep learning library [pytorch](https://pytorch.org) is not installed. 
I did not include the package in the `requirements.txt` because of GPU support.
Just run `pip install torch` and try again or follow the official installation [here](https://pytorch.org/get-started/locally/)
Make sure you downloaded the model parameters, linked in the README and you should be ready to go.

**Reply by mrd33p** [reply]
More errors, after installing torch.

`(base) PS C:\Users\DELL\Downloads\LaTeX-OCR-main> python gui.py
Traceback (most recent call last):
  File "gui.py", line 16, in <module>
    import pix2tex
  File "C:\Users\DELL\Downloads\LaTeX-OCR-main\pix2tex.py", line 1, in <module>
    from dataset.dataset import test_transform
  File "C:\Users\DELL\Downloads\LaTeX-OCR-main\dataset\__init__.py", line 4, in <module>
    import dataset.render
  File "C:\Users\DELL\Downloads\LaTeX-OCR-main\dataset\render.py", line 11, in <module>
    import cv2
  File "e:\ProgramData\Anaconda3\lib\site-packages\cv2\__init__.py", line 9, in <module>
    from .cv2 import _registerMatType
ImportError: cannot import name '_registerMatType' from 'cv2.cv2' (e:\ProgramData\Anaconda3\lib\site-packages\cv2\cv2.cp37-win_amd64.pyd)
(base) PS C:\Users\DELL\Downloads\LaTeX-OCR-main>`

Here's what I did,

1. Installed torch using `pip install torch`
2. ran command `pip install -r requirements.txt` again, it showed some errors
3. downloaded `weights.pth` and `image_resizer.pth` and pasted in `checkpoints` folder
4. tried running `gui.py` and then ended up with the error above. 

what I am missing?

**Reply by lukas-blecher** [reply]
That's weird. Can you show me the  error?
I've since figured out what the problem was (#72) and fixed it.
In this comment: https://github.com/lukas-blecher/LaTeX-OCR/issues/72#issuecomment-1003180696 jave performed similar steps as you and had success.

**Comment by mrd33p** [comment]
The earlier error was showing Visual Studio C++ is not installed, then I installed VS and ran the command `pip install -r requirements.txt` again, which got executed properly, but then when I tried to run `python gui.py`, it showed the following error
`(base) PS C:\Users\DELL\Downloads\LaTeX-OCR-main> python gui.py
Traceback (most recent call last):
  File "gui.py", line 16, in <module>
    import pix2tex
  File "C:\Users\DELL\Downloads\LaTeX-OCR-main\pix2tex.py", line 1, in <module>
    from dataset.dataset import test_transform
  File "C:\Users\DELL\Downloads\LaTeX-OCR-main\dataset\__init__.py", line 4, in <module>
    import dataset.render
  File "C:\Users\DELL\Downloads\LaTeX-OCR-main\dataset\render.py", line 11, in <module>
    import cv2
  File "e:\ProgramData\Anaconda3\lib\site-packages\cv2\__init__.py", line 9, in <module>
    from .cv2 import _registerMatType
ImportError: cannot import name '_registerMatType' from 'cv2.cv2' (e:\ProgramData\Anaconda3\lib\site-packages\cv2\cv2.cp37-win_amd64.pyd)
(base) PS C:\Users\DELL\Downloads\LaTeX-OCR-main>`

What's wrong this time?

**Reply by mrd33p** [reply]
I got it working; I just uninstalled and reinstalled opencv-python using `pip uninstall opencv-python` and then reinstalled `pip install opencv-python`. Just for information if anyone else needs it.
But then it said 'wandb' is not found, so I installed it as well, using `pip install wandb`. It works now, but I am not getting the expected results. 

I was expecting it to work like an equation in word. It shows errors like 'Missing or unrecognized delimiter for \Big' etc., 
![errror](https://user-images.githubusercontent.com/85106864/147849223-82725b4f-3dcf-4b2a-8cb2-3b8948389bb1.jpg)

and if I paste it in word as latex, it doesn't recognize the code. Does it do that? The picture below shows unrecognizable latex pasted in word
![word error](https://user-images.githubusercontent.com/85106864/147849231-04940c2d-8053-402f-9f1d-fc0f6eb44f06.jpg)

**Reply by lukas-blecher** [reply]
Sorry that the model does not work as expected. Can I take a look at these snips?
There are still some shortcomings with the current method mainly for multi line equations.

The reason you see the error message is that it is not actually checked if the output is valid latex code and the mathjax compiler in the gui can't show parital correct equations.

**Reply by mrd33p** [reply]
No need to feel sorry, your work is amazing and should be promoted in the scientific/education community. 
I got it working for some equations.

Anyway, will training improve the results? I tried training the models with your data but faced some errors again, 
I put data which is downloaded from g drive in folder `LaTeX-OCR-main\td` see image 

 ![image](https://user-images.githubusercontent.com/85106864/147868316-526f2a8e-c7f1-479b-95e9-88941f94ea67.png)

then run this command in `dataset` folder for training 
`python dataset.py --equations C:\Users\DELL\Downloads\LaTeX-OCR-main\td\math.txt --images C:\Users\DELL\Downloads\LaTeX-OCR-main\td --tokenizer dataset/tokenizer.json --out dataset.pkl`

but it showed this error 
`(base) PS C:\Users\DELL\Downloads\LaTeX-OCR-main\dataset> python dataset.py --equations C:\Users\DELL\Downloads\LaTeX-OCR-main\td\math.txt --images C:\Users\DELL\Downloads\LaTeX-OCR-main\td --tokenizer dataset/tokenizer.json --out dataset.pkl
Generate dataset
Traceback (most recent call last):
  File "dataset.py", line 247, in <module>
    Im2LatexDataset(args.equations, args.images, args.tokenizer).save(args.out)
  File "dataset.py", line 86, in __init__
    self.tokenizer = PreTrainedTokenizerFast(tokenizer_file=tokenizer)
  File "e:\ProgramData\Anaconda3\lib\site-packages\transformers\tokenization_utils_fast.py", line 86, in __init__
    fast_tokenizer = TokenizerFast.from_file(fast_tokenizer_file)
Exception: The system cannot find the path specified. (os error 3)
(base) PS C:\Users\DELL\Downloads\LaTeX-OCR-main\dataset>
`

I am confused about whether I gave the address for images and text file correctly. What's causing this error?

**Reply by lukas-blecher** [reply]
Training can improve it (only try if you have a nvidia gpu and torch with cuda support) but I think more training data is more important at the moment. Anyway to solve your problem:
1. You have to run the command from the root of the project, not the dataset directory. 
2. The images argument must point to one of train, test or val. Change the out argument accordingly. `math.txt` stays the same.

That should solve the problem.

**Comment by aburousan** [comment]
hey can someone please tell me how to use the LaTexOCR window in ubuntu?
like shown in the gif?
